internal/form/submit/service_test.go

```diff
@@ -0,0 +1,257 @@
+		validate    func(t *testing.T, p Params, got response.FormResponse, errs []error)
+	}
+
+	formID := uuid.New()
+	userID := uuid.New()
+	q1, q2 := uuid.New(), uuid.New()
+	unknown := uuid.New()
+
+	testCases := []testCase{
+		{
+			name: "CreateOrUpdate called with mapped types if all valid",
+			params: Params{
+				formID:  formID,
+				userID:  userID,
+				qs:      []question.Answerable{makeQ(q1, question.QuestionTypeShortText, nil), makeQ(q2, question.QuestionTypeLongText, nil)},
+				answers: []shared.AnswerParam{ans(q1, "hello"), ans(q2, "world")},
+			},
+			setup: func(t *testing.T, p *Params) context.Context {
+				qm := mocks.NewMockQuestionStore(t)
+				rm := mocks.NewMockFormResponseStore(t)
+
+				qm.EXPECT().
+					ListByFormID(mock.Anything, p.formID).
+					Return(p.qs, nil).Once()
+
+				respID := uuid.New()
+				rm.EXPECT().
+					CreateOrUpdate(
+						mock.Anything,
+						p.formID,
+						p.userID,
+						p.answers,
+						[]response.QuestionType{
+							response.QuestionType(question.QuestionTypeShortText),
+							response.QuestionType(question.QuestionTypeLongText),
+						},
+					).
+					Return(response.FormResponse{ID: respID, FormID: p.formID, SubmittedBy: p.userID}, nil).
+					Once()
+
+				p.qStore = qm
+				p.rStore = rm
+				p.service = submit.NewService(zap.NewNop(), qm, rm)
+				return context.Background()
+			},
+			validate: func(t *testing.T, p Params, got response.FormResponse, errs []error) {
+				require.Nil(t, errs)
+				require.Equal(t, p.formID, got.FormID)
+				require.Equal(t, p.userID, got.SubmittedBy)
+				require.NotEqual(t, uuid.Nil, got.ID)
+			},
+		},
+		{
+			name: "Not call CreateOrUpdate if there are some invalid answers",
+			params: Params{
+				formID:  formID,
+				userID:  userID,
+				qs:      []question.Answerable{makeQ(q1, question.QuestionTypeShortText, nil), makeQ(q2, question.QuestionTypeLongText, errors.New("invalid"))},
+				answers: []shared.AnswerParam{ans(q1, "ok"), ans(q2, "bad")},
+			},
+			expectedErr: true,
+			setup: func(t *testing.T, p *Params) context.Context {
+				qm := mocks.NewMockQuestionStore(t)
+				rm := mocks.NewMockFormResponseStore(t)
+
+				qm.EXPECT().
+					ListByFormID(mock.Anything, p.formID).
+					Return(p.qs, nil).Once()
+
+				p.qStore = qm
+				p.rStore = rm
+				p.service = submit.NewService(zap.NewNop(), qm, rm)
+				return context.Background()
+			},
+			validate: func(t *testing.T, p Params, got response.FormResponse, errs []error) {
+				require.NotNil(t, errs)
+				require.Len(t, errs, 1)
+				require.Equal(t, uuid.Nil, got.ID)
+				p.rStore.AssertNotCalled(t, "CreateOrUpdate", mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything)
+			},
+		},
+		{
+			name: "Not call CreateOrUpdate if answer refers to unknown question ",
+			params: Params{
+				formID:  formID,
+				userID:  userID,
+				qs:      []question.Answerable{makeQ(q1, question.QuestionTypeShortText, nil)},
```

Comment:
Consider to rename it to `questions` for better readability
```suggestion
				questions:      []question.Answerable{makeQ(q1, question.QuestionTypeShortText, nil)},
```